extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergCatalogAdapter.java

```diff
@@ -0,0 +1,259 @@
+import io.deephaven.engine.table.impl.locations.impl.StandaloneTableKey;
+import io.deephaven.engine.table.impl.locations.impl.TableLocationKeyFinder;
+import io.deephaven.engine.table.impl.locations.util.TableDataRefreshService;
+import io.deephaven.engine.table.impl.sources.regioned.RegionedTableComponentFactoryImpl;
+import io.deephaven.engine.updategraph.UpdateSourceRegistrar;
+import io.deephaven.iceberg.layout.IcebergFlatLayout;
+import io.deephaven.iceberg.layout.IcebergKeyValuePartitionedLayout;
+import io.deephaven.iceberg.location.IcebergTableLocationFactory;
+import io.deephaven.iceberg.location.IcebergTableLocationKey;
+import org.apache.iceberg.PartitionField;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.catalog.Catalog;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+import org.jetbrains.annotations.NotNull;
+
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class IcebergCatalogAdapter {
+    private final Catalog catalog;
+    private final FileIO fileIO;
+    private final IcebergInstructions instructions;
+
+    private final WeakHashMap<PartitionAwareSourceTable, AbstractTableLocationProvider> tableLocationProviders;
+
+    /**
+     * Construct an IcebergCatalogAdapter given a set of configurable instructions.
+     */
+    @SuppressWarnings("unused")
+    IcebergCatalogAdapter(
+            @NotNull final Catalog catalog,
+            @NotNull final FileIO fileIO) {
+        this(catalog, fileIO, IcebergInstructions.builder().build());
+    }
+
+    /**
+     * Construct an IcebergCatalogAdapter given a set of configurable instructions.
+     */
+    IcebergCatalogAdapter(
+            @NotNull final Catalog catalog,
+            @NotNull final FileIO fileIO,
+            @NotNull final IcebergInstructions instructions) {
+        this.catalog = catalog;
+        this.fileIO = fileIO;
+        this.instructions = instructions;
+        tableLocationProviders = new WeakHashMap<>();
+    }
+
+    static TableDefinition fromSchema(
+            final Schema schema,
+            final PartitionSpec partitionSpec,
+            final IcebergInstructions instructions) {
+        final Map<String, String> renameMap = instructions.columnRenameMap();
+
+        final Set<String> partitionNames =
+                partitionSpec.fields().stream()
+                        .map(PartitionField::name)
+                        .map(colName -> renameMap.getOrDefault(colName, colName))
+                        .collect(Collectors.toSet());
+
+        final List<ColumnDefinition<?>> columns = new ArrayList<>();
+
+        for (final Types.NestedField field : schema.columns()) {
```

Comment:
Is it possible to have a derived partitioning column that is *not* in the schema? If so, should we be adding it?